datafusion/physical-expr/src/aggregate/sum.rs

```diff
@@ -262,98 +249,83 @@ fn sum_decimal_with_diff_scale(
-                sum_decimal_with_diff_scale(v2, v1, max_precision, s2, s1)
+
+            match (lhs, rhs) {
+                (
+                    ScalarValue::Decimal128(v1, _, _),
+                    ScalarValue::Decimal128(v2, _, _),
+                ) => {
+                    Ok(if s1.eq(&s2) {
+                        // s1 = s2
+                        sum_decimal(v1, v2, &max_precision, &s1)
+                    } else if s1.gt(&s2) {
+                        // s1 > s2
+                        sum_decimal_with_diff_scale(v1, v2, &max_precision, &s1, &s2)
+                    } else {
+                        // s1 < s2
+                        sum_decimal_with_diff_scale(v2, v1, &max_precision, &s2, &s1)
+                    })
+                }
+                _ => Err(DataFusionError::Internal(
+                    "Internal state error on sum decimals ".to_string(),
+                )),
             }
         }
-        // float64 coerces everything to f64
-        (ScalarValue::Float64(lhs), ScalarValue::Float64(rhs)) => {
-            typed_sum!(lhs, rhs, Float64, f64)
-        }
-        (ScalarValue::Float64(lhs), ScalarValue::Float32(rhs)) => {
-            typed_sum!(lhs, rhs, Float64, f64)
-        }
-        (ScalarValue::Float64(lhs), ScalarValue::Int64(rhs)) => {
-            typed_sum!(lhs, rhs, Float64, f64)
-        }
-        (ScalarValue::Float64(lhs), ScalarValue::Int32(rhs)) => {
-            typed_sum!(lhs, rhs, Float64, f64)
-        }
-        (ScalarValue::Float64(lhs), ScalarValue::Int16(rhs)) => {
-            typed_sum!(lhs, rhs, Float64, f64)
-        }
-        (ScalarValue::Float64(lhs), ScalarValue::Int8(rhs)) => {
-            typed_sum!(lhs, rhs, Float64, f64)
+        (DataType::Float64, _) | (_, DataType::Float64) => {
+            let data: ArrayRef =
+                union_arrays!(lhs, rhs, &DataType::Float64, Float64Array, "f64");
+            sum_batch(&data, &arrow::datatypes::DataType::Float64)
         }
-        (ScalarValue::Float64(lhs), ScalarValue::UInt64(rhs)) => {
-            typed_sum!(lhs, rhs, Float64, f64)
-        }
-        (ScalarValue::Float64(lhs), ScalarValue::UInt32(rhs)) => {
-            typed_sum!(lhs, rhs, Float64, f64)
-        }
-        (ScalarValue::Float64(lhs), ScalarValue::UInt16(rhs)) => {
-            typed_sum!(lhs, rhs, Float64, f64)
-        }
-        (ScalarValue::Float64(lhs), ScalarValue::UInt8(rhs)) => {
-            typed_sum!(lhs, rhs, Float64, f64)
-        }
-        // float32 has no cast
-        (ScalarValue::Float32(lhs), ScalarValue::Float32(rhs)) => {
-            typed_sum!(lhs, rhs, Float32, f32)
-        }
-        // u64 coerces u* to u64
-        (ScalarValue::UInt64(lhs), ScalarValue::UInt64(rhs)) => {
-            typed_sum!(lhs, rhs, UInt64, u64)
+        (DataType::Float32, _) | (_, DataType::Float32) => {
+            let data: ArrayRef =
+                union_arrays!(lhs, rhs, &DataType::Float32, Float32Array, "f32");
```

Comment:
this is an interesting idea, but I suspect the performance will be fairly low (as it creates arrays for each value 🤔 )

I wonder if we could move the `sum` logic into `scalar.rs` and instead add some sort of coertion logic 

Not sure.

Reply:
do you mean add a sum function to scalar value struct?

like 
```
 sum(self, other: ScalarValue) -> ScalarValue
```

Reply:
Something like that -- yes -- given that is effectively what this code is doing. 

@realno  added just such code here https://github.com/apache/arrow-datafusion/pull/1525 though I was trying to avoid it https://github.com/apache/arrow-datafusion/pull/1550

However, if it is truly required then having one copy of it seems better than several 🤔

Reply:
Went through those PRs. Also currently we have negate function on scalar value level. 
```
    pub fn arithmetic_negate(&self) -> Self {
```
Probably we can extend this trend and add other common math functions to Scalar Value too like sum, minus, multiply, divide?

Reply:
I am torn -- basically I worry that adding the common math functions to `ScalarValue` will result in them being used more (and they are very slow). See more discussion on  https://github.com/apache/arrow-datafusion/pull/1525#pullrequestreview-847147545

However, if the alternative is a bunch of replicated code over the codebase, consolidating that all into ScalarValue seems like a much better outcome

Reply:
okay, I found its already implemented coercion for sum result type

```
        AggregateFunction::Sum => sum_return_type(&coerced_data_types[0]),
```

I'll check, probably coercing the resulting sum will allow to get rid of coercion in the `sum.rs`

Reply:
thanks@

Reply:
> implementing the add function(in scalar.rs or any other place) leads to boilerplate code, because there is no concise way how to get ScalarValue underlying value.

I think this is the best we can do -- I don't (really) mind boiler plate code, but there seems to be too many copies of almost the same boiler plate code.

I agree using `arrow::compute::sum` is likely not the right thing to do

Reply:
Right, I'm still thinking how we can do that. I checked arrow-rs but they dont have the same model with enum + option.
Even if we implement once the add function with boilerplate code, and then we need to implement minus function or any other function requiring direct actions with scalarValue underlying values, we will have to add boilerplate code again. The reason being we can't access underlying value without bunch of matches....

I think to make things easier we need to implement a function to return inner value somehow. Kind of 
`scalarValue.unwrap()`

@alamb perhaps you have better thoughts?

Reply:
I think we had a bunch of discussion on a related topic (how to handle constants and arrays similarly) on https://github.com/apache/arrow-datafusion/issues/1248

Maybe that will provide some insight

Reply:
I came up to idea like that, it should put the boilerplate once. And then use external functions over values.

```
use std::ops::Sub;
use std::ops::Add;
use std::fmt::Debug;
use std::any::Any;

#[derive(Debug, Copy, Clone)]
enum DataType {
    Int32(Option<i32>),
    Float64(Option<f64>)
}

// boilerplate comes here
macro_rules! op {
    ($ARG1: expr, $ARG2: expr, $FUNC: block) => {{
        let res = match ($ARG1, $ARG2) {
            (DataType::Int32(Some(v1)), DataType::Float64(Some(v2))) => 
            DataType::Float64(Some($FUNC(v1 as f64, v2 as f64))),
            _ => panic!("123")
        };
        
        res
    }};
}

fn sum<T:Add<Output = T> + Copy> (num1: T, num2: T) -> T {
    return num1 + num2;
}

fn minus<T:Sub<Output = T> + Copy> (num1: T, num2: T) -> T {
    return num1 - num2;
}

fn main() {
    let i_32 = DataType::Int32(Some(1));
    let f_64 = DataType::Float64(Some(2.0));

    let s = op!(i_32, f_64, {sum});
    dbg!(&s);
    
    let m = op!(i_32, f_64, {minus});
    dbg!(&m);
}
```

Reply:
https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=a513ccd564230b3815d635c515236f5f

Reply:
What about just cut all this coercions logic? I've investigate all the occurrences of `sum()`, it's only used to accumulate aggregator state in `sum`, `sum_distinct` and `average` where the operand's type of `sum()` is deterministic. And `sum()` is an internal function (`pub(crate)`), API change of this function is acceptable.

I try to remove all the match arms that with different operand types and only fail two cases (`sum_distinct_i32_with_nulls` and `sum_distinct_u32_with_nulls`). I think this is acceptable. And I find the min/max calculator already applied this.

About how to achieve calculate operator over different types, I think we can extract our coercion rule to something like
```rust
fn coercion(lhs: DataType, rhs: DataType) -> DataType {}
```
And cast both operands to the result type before calculation.